core/os/device/remotessh/device.go

```diff
@@ -124,24 +132,132 @@ func (b *binding) newPooledSession() (*pooledSession, error) {
 	}, nil
 }
 
+// Interface check
 var _ Device = &binding{}
 
-// Devices returns the list of reachable SSH devices.
-func Devices(ctx context.Context, configuration io.Reader) ([]bind.Device, error) {
-	configurations, err := ReadConfigurations(configuration)
-	if err != nil {
+var (
+	// Registry of all the discovered devices.
+	registry = bind.NewRegistry()
+
+	// cache is a map of device names to fully resolved bindings.
+	cache      = map[string]*binding{}
+	cacheMutex sync.Mutex // Guards cache.
+)
+
+// Monitor updates the registry with devices that are added and removed at the
+// specified interval. Monitor returns once the context is cancelled.
+func Monitor(ctx context.Context, r *bind.Registry, interval time.Duration, conf func() ([]io.ReadCloser, error)) error {
+	unlisten := registry.Listen(bind.NewDeviceListener(r.AddDevice, r.RemoveDevice))
+	defer unlisten()
+
+	for _, d := range registry.Devices() {
+		r.AddDevice(ctx, d)
+	}
+
+	var lastErrorPrinted time.Time
+
+	for {
+		if err := func() error {
+			rcs, err := conf()
+			if err != nil {
+				return err
+			}
+			defer func() {
+				for _, rc := range rcs {
+					rc.Close()
+				}
+			}()
+			configs := []Configuration{}
+			for _, rc := range rcs {
+				configurations, err := ReadConfigurations(rc)
```

Comment:
Nit: Looks like there are some shared code here and in `remotessh.Devices()`, they all read a list of configurations and then call `scanDevices`